atef/widgets/archive_viewer.py

```diff
@@ -0,0 +1,619 @@
+from qtpy.QtWidgets import QWidget
+
+from atef.widgets.core import DesignerDisplay
+
+logger = logging.getLogger(__name__)
+archive_viewer_singleton = None
+ARCHIVER_URLS = ['http://pscaa01.slac.stanford.edu',
+                 'http://pscaa02.slac.stanford.edu']
+symbol_map = {'None': None, 'circle': 'o', 'square': 's',
+              'cross': '+', 'star': 'star'}
+style_map = {'solid': Qt.SolidLine,
+             'dash': Qt.DashLine, 'dot': Qt.DotLine}
+color_cycle = itertools.cycle(
+    [QtGui.QColor('red'), QtGui.QColor('blue'),
+     QtGui.QColor('green'), QtGui.QColor('white')]
+)
+
+
+def get_archive_viewer() -> ArchiverViewerWidget:
+    """
+    Only allow one viewer to be open at a time.
+    Makes it unambiguous where to send PV's to.
+
+    Returns
+    -------
+    ArchiveViewerWidget
+        the widget instance
+    """
+    global archive_viewer_singleton
+    if archive_viewer_singleton is None:
+        archive_viewer_singleton = ArchiverViewerWidget()
+    return archive_viewer_singleton
+
+
+def get_reachable_url(urls: List[str]) -> str:
+    """
+    Get valid archiver URLS from the urls
+    Looks only for an response code below 400, as a proxy ping test.
+    Ideally the urls provided by the env var work, but for now we
+    err on the side of caution
+
+    Returns
+    -------
+    str
+        a responsive url
+    """
+    for url in urls:
+        try:
+            _ = urllib.request.urlopen(url, timeout=1)
+        except urllib.error.URLError as e:
+            if isinstance(e, urllib.error.HTTPError):
+                # HTTPError subclasses URLError, separates code
+                errno = e.code
+            elif e.reason.errno:
+                errno = e.reason.errno
+            else:
+                errno = -1
+
+            if 0 < errno < 400:
+                # timeouts give socket.timeout and have no errno
+                # here we only care if the server exists
+                # connection refused is ok
+                return url
+
+
+def _success_decorator(method):
+    """
+    Decorator for qt methods that want to return True if successful and
+    False otherwise.  Wraps the method in a simple try-except.
+    """
+    def _dec(self, *args, **kwargs):
+        try:
+            method(self, *args, **kwargs)
+            return True
+        except Exception as e:
+            logger.debug(e)
+            return False
+    return _dec
+
+
+class ArchiverError(Exception):
+    """ Archiver related exceptions """
+    ...
+
+
+class ArchiverViewerWidget(DesignerDisplay, QWidget):
+    """
+    Archiver time plot viewer
+    """
+    filename: ClassVar[str] = 'archive_viewer_widget.ui'
+
+    time_plot: PyDMArchiverTimePlot
+    button_month: QtWidgets.QPushButton
+    button_week: QtWidgets.QPushButton
+    button_day: QtWidgets.QPushButton
+    input_field: QtWidgets.QLineEdit
+    curve_list: QtWidgets.QTableView
+    redraw_button: QtWidgets.QPushButton
```

Comment:
A "clear curves" button might also be nice, for when you want to start from scratch